enterprise/tailnet/pgcoord.go

```diff
@@ -683,21 +899,116 @@ func (q *querier) newConn(c *connIO) {
 		return
 	}
 	cm.count++
-	q.conns[c] = struct{}{}
-	go q.cleanupConn(c)
+	q.conns[c.UniqueID()] = c
+	go q.waitCleanupConn(c)
 }
 
-func (q *querier) cleanupConn(c *connIO) {
-	<-c.ctx.Done()
+func (q *querier) newClientSubscription(c agpl.Queue, agentID uuid.UUID) {
 	q.mu.Lock()
 	defer q.mu.Unlock()
-	delete(q.conns, c)
+
+	if _, ok := q.clientSubscriptions[c.UniqueID()]; !ok {
+		q.clientSubscriptions[c.UniqueID()] = map[uuid.UUID]struct{}{}
+	}
+
 	mk := mKey{
-		agent: c.agent,
-		// if client is Nil, this is an agent connection, and it wants the mappings for all the clients of itself
-		clientsOfAgent: c.client == uuid.Nil,
+		agent: agentID,
+		kind:  agpl.QueueKindClient,
+	}
+	cm, ok := q.mappers[mk]
+	if !ok {
+		ctx, cancel := context.WithCancel(q.ctx)
+		mpr := newMapper(ctx, q.logger, mk, q.heartbeats)
+		cm = &countedMapper{
+			mapper: mpr,
+			count:  0,
+			cancel: cancel,
+		}
+		q.mappers[mk] = cm
+		// we don't have any mapping state for this key yet
+		q.workQ.enqueue(mk)
+	}
+	if err := sendCtx(cm.ctx, cm.add, c); err != nil {
+		return
+	}
+	q.clientSubscriptions[c.UniqueID()][agentID] = struct{}{}
+	cm.count++
+}
+
+func (q *querier) removeClientSubscription(c agpl.Queue, agentID uuid.UUID) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	mk := mKey{
+		agent: agentID,
+		kind:  agpl.QueueKindClient,
 	}
 	cm := q.mappers[mk]
+	if err := sendCtx(cm.ctx, cm.del, c); err != nil {
+		return
+	}
+	delete(q.clientSubscriptions[c.UniqueID()], agentID)
+	cm.count--
+	if cm.count == 0 {
+		cm.cancel()
+		delete(q.mappers, mk)
+	}
+}
+
+func (q *querier) newClientConn(c agpl.Queue) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	if !q.healthy {
+		err := c.Close()
+		q.logger.Info(q.ctx, "closed incoming connection while unhealthy",
+			slog.Error(err),
+			slog.F("client_id", c.UniqueID()),
+		)
+		return
+	}
+
+	q.conns[c.UniqueID()] = c
+	go q.waitCleanupConn(c)
+}
+
+func (q *querier) waitCleanupConn(c agpl.Queue) {
+	<-c.Done()
+	q.cleanupConn(c)
+}
+
+func (q *querier) cleanupConn(c agpl.Queue) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	delete(q.conns, c.UniqueID())
+
+	// Iterate over all subscriptions and remove them from the mappers.
+	for agentID := range q.clientSubscriptions[c.UniqueID()] {
+		mk := mKey{
+			agent: agentID,
+			kind:  c.Kind(),
+		}
+		cm, ok := q.mappers[mk]
+		if ok {
```

Comment:
the fact that you're checking `ok` here smells funny and indicates that we're not keeping the data structures in sync.

It should be an invariant in this code that a client subscription in the `q.clientSubscriptions` map implies that a corresponding mapper exists.  They are always added together and always removed together while holding the mutex.

Reply:
How do you think we should handle a case where the mapper doesn't exist. Panic?

Reply:
yeah, panic, or at the very least, drop a `Critical` log

Reply:
I found a pretty easy way to trigger this invariant. It's possible for `(*querier).cleanupConn` to race with a call to `removeClientSubscription`, such that `removeClientSubscription` is trying to remove a subscription that no longer exists. I'm not too sure how to prevent this, except by ignoring duplicates.

Reply:
The invariant is that a client subscription exists in `q.clientSubscriptions` AND in the mappers, or neither.  It can never be in one but not the other when the mutex is unlocked.

You're right though, that in this design, it's possible for `removeClientSubscription` to race with `cleanupConn` such that you can try to remove a subscription that was already cleaned up.  What you've done is fine and follows the invariant --- if the subscription is not in `q.clientSubscriptions` that means it's already cleaned up from the mappers too, and `removeClientSubscription` can return early.